core/blockchain_pruner.go

```diff
@@ -0,0 +1,256 @@
+	if bp.batchWriter != nil {
+		err := bp.batchWriter.Delete(key)
+		if err != nil {
+			if isUnrecoverableErr(err) {
+				fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+				os.Exit(1)
+			}
+			return err
+		}
+
+		if bp.batchWriter.ValueSize() >= maxMemoryUsage {
+			err = bp.batchWriter.Write()
+			if err != nil {
+				if isUnrecoverableErr(err) {
+					fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+					os.Exit(1)
+				}
+				return err
+			}
+
+			bp.batchWriter.Reset()
+		}
+		return nil
+	} else {
+		return errors.New("delete when batchWriter not init")
+	}
+}
+
+func (bp *blockchainPruner) endPruneBeaconChain() error {
+	if bp.batchWriter != nil {
+		err := bp.batchWriter.Write()
+		if err != nil {
+			if isUnrecoverableErr(err) {
+				fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+				os.Exit(1)
+			}
+			return err
+		}
+
+		bp.batchWriter = nil
+		return nil
+	} else {
+		return errors.New("end when batchWriter not init")
+	}
+}
+
+func (bp *blockchainPruner) runWhenNotPruning(cb func()) {
+	// check is start
+	if atomic.CompareAndSwapInt32(&bp.isPruneBeaconChainRunning, 0, 1) {
+		defer atomic.StoreInt32(&bp.isPruneBeaconChainRunning, 0)
+
+		cb()
+	}
+}
+
+func (bp *blockchainPruner) StartPruneBeaconChain(db ethdb.Database, maxBlockNum uint64, maxEpoch *big.Int) {
+	bp.runWhenNotPruning(func() {
+		err := bp.initPruneBeaconChain(db)
+		if err != nil {
+			utils.Logger().Info().Err(err).Msg("pruneBeaconChain init error")
+			return
+		}
+
+		defer func() {
+			err := bp.endPruneBeaconChain()
+			if err != nil {
+				utils.Logger().Info().Err(err).Msg("pruneBeaconChain end error")
+				return
+			}
+
+			utils.Logger().Info().
+				Uint64("maxBlockNum", maxBlockNum).
+				Uint64("maxEpoch", maxEpoch.Uint64()).
+				Int("deletedBlockCount", bp.deletedBlockCount).
+				Int("deletedValidatorSnapshot", bp.deletedValidatorSnapshot).
+				Int("skipValidatorSnapshot", bp.skipValidatorSnapshot).
+				Dur("cost", time.Now().Sub(bp.startTime)).
+				Msg("pruneBeaconChain delete block success")
+
+			if rand.Intn(compactProportion) == 1 {
+				startTime := time.Now()
+				// probability of 1 in 1000 blocks to Compact, It consumes a lot of IO
+				for _, rangeStartEnd := range bp.deleteRange {
+					err := db.Compact(rangeStartEnd[0], rangeStartEnd[1])
+					if err != nil {
+						if isUnrecoverableErr(err) {
+							fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+							os.Exit(1)
+						}
+					}
+				}
+				utils.Logger().Info().
+					Uint64("maxBlockNum", maxBlockNum).
+					Uint64("maxEpoch", maxEpoch.Uint64()).
+					Dur("cost", time.Now().Sub(startTime)).
+					Msg("pruneBeaconChain compact db success")
+			}
+		}()
+
+		bp.pruneBeaconChainBlock(db, maxBlockNum)
+		bp.pruneBeaconChainValidatorSnapshot(db, maxEpoch)
+	})
+}
+
+func (bp *blockchainPruner) pruneBeaconChainBlock(reader ethdb.Database, maxBlockNum uint64) {
```

Comment:
Looks like this will iterate through the whole beaconchain for each start. Any more elegant solution?

Reply:
It does not directly iterate the entire beaconchain, it will quickly find the first block that has not been deleted (same as Get), and then read sequentially backwards, This is a feature of LevelDB